File: apps/desktop/electron/reminders.ts
/**
 * 提醒规则引擎（M4）—— 四条规则 + 去重 + 贪睡静默
 *
 * 规则（每次快照变化 / 每分钟 tick 时评估）：
 * 1. heartbeat  连续专注 50 分钟（同一 doing 任务 + 主屏计入）→ 休息提醒
 * 2. drift      有 doing 任务，但过去 5 分钟主屏计入时长 < 30s（分心）→ 轻推
 * 3. switch_storm 5 分钟内前台切换 ≥ 15 次（高频跳窗）→ 平复提醒
 * 4. idle_back  空闲 ≥ 10 分钟后恢复输入 → 欢迎/确认回来
 *
 * 通用约束：
 * - 每种规则 10 分钟内最多触发一次（同 key 去重）
 * - 用户点「稍后提醒」→ 该规则静默 snoozeMs（默认 20 分钟）
 * - overlay 可见时不再触发新提醒
 */
import { onSnapshotChange, onSessionEnded, isIdle } from './tracker';
import { getCurrentTask } from './db';
import { showOverlay, hideOverlay, isOverlayVisible, type OverlayPayload } from './overlay';

/* ── 配置（M5 设置页可改）────────────────────────────────── */
export interface ReminderConfig {
  heartbeatFocusMs: number;      // 专注多久提醒（默认 50min）
  driftWindowMs: number;         // 分心检测窗口（默认 5min）
  driftMinCountedMs: number;     // 窗口内最少计入时长（默认 30s）
  stormWindowMs: number;         // 切换风暴窗口（默认 5min）
  stormSwitchThreshold: number;  // 切换次数阈值（默认 15）
  idleBackMinMs: number;         // 空闲多久恢复算 idle_back（默认 10min）
  perRuleCooldownMs: number;     // 同规则去重间隔（默认 10min）
  snoozeMs: number;              // 「稍后提醒」静默时长（默认 20min）
}

export const DEFAULT_CONFIG: ReminderConfig = {
  heartbeatFocusMs: 50 * 60_000,
  driftWindowMs: 5 * 60_000,
  driftMinCountedMs: 30_000,
  stormWindowMs: 5 * 60_000,
  stormSwitchThreshold: 15,
  idleBackMinMs: 10 * 60_000,
  perRuleCooldownMs: 10 * 60_000,
  snoozeMs: 20 * 60_000,
};

type RuleKey = 'heartbeat' | 'drift' | 'switch_storm' | 'idle_back';

const RULE_TITLE: Record<RuleKey, string> = {
  heartbeat: '连续专注了一段时间',
  drift: '好像走神了？',
  switch_storm: '切换有点频繁',
  idle_back: '欢迎回来',
};

/* ── 运行时状态 ──────────────────────────────────────────── */

let cfg: ReminderConfig = { ...DEFAULT_CONFIG };
let enabled = true;
let paused = false; // 托盘「暂停提醒」

// 去重 / 静默
const lastFiredAt = new Map<RuleKey, number>();
const snoozedUntil = new Map<RuleKey, number>();

// heartbeat：当前 doing 任务累计的连续计入时长（会话事件累加，切任务/空闲即清零）
let continuousFocusMs = 0;
let continuousTaskId: number | null = null;

// switch 计数环形记录（时间窗口用）
const switchTimes: number[] = [];

// drift：窗口内计入时长（滚动求和）
const countedSegments: Array<{ at: number; ms: number }> = [];

// idle_back：进入空闲的时刻与时长
let idleStartedAt = 0;

let listeners: Array<(r: OverlayPayload) => void> = [];
export function onReminderFired(fn: (r: OverlayPayload) => void): void {
  listeners.push(fn);
}

/* ── 对外控制（M5 设置 / 托盘菜单用）────────────────────── */

export function startReminders(config?: Partial<ReminderConfig>): void {
  if (config) cfg = { ...cfg, ...config };
  wireTracking();
}

export function stopReminders(): void {
  enabled = false;
}

export function setRemindersPaused(p: boolean): void {
  paused = p;
  if (p) hideOverlay();
}

export function snoozeRule(rule: RuleKey, ms = cfg.snoozeMs): void {
  snoozedUntil.set(rule, Date.now() + ms);
}

export function snoozeAll(ms = cfg.snoozeMs): void {
  (['heartbeat', 'drift', 'switch_storm', 'idle_back'] as RuleKey[]).forEach((k) => snoozeRule(k, ms));
}

/** 渲染层点「稍后提醒」回调 */
export function handleOverlaySnooze(): void {
  // 找到当前可见的提醒对应规则：简化为全部静默（一次只有一个浮层）
  snoozeAll();
  hideOverlay();
}

/* ── 内部实现 ────────────────────────────────────────────── */

function canFire(rule: RuleKey, now: number): boolean {
  if (!enabled || paused) return false;
  if (isOverlayVisible()) return false;
  const snoozed = snoozedUntil.get(rule) ?? 0;
  if (now < snoozed) return false;
  const last = lastFiredAt.get(rule) ?? 0;
  if (now - last < cfg.perRuleCooldownMs) return false;
  return true;
}

function fire(rule: RuleKey, body: string): void {
  const now = Date.now();
  lastFiredAt.set(rule, now);
  const payload: OverlayPayload = {
    kind: rule,
    title: RULE_TITLE[rule],
    body,
  };
  void showOverlay(payload);
  for (const fn of listeners) {
    try { fn(payload); } catch { /* listener error 不影响主流程 */ }
  }
}

let wired = false;

function wireTracking(): void {
  if (wired) return;
  wired = true;

  // 会话结束：累计 heartbeat 专注 / drift 计入 / switch 计数
  onSessionEnded((s) => {
    const now = s.endedAt;

    // switch_storm：所有前台会话都算一次切换（含不计入的副屏）
    switchTimes.push(now);
    while (switchTimes.length > 0 && now - switchTimes[0] > cfg.stormWindowMs) switchTimes.shift();

    if (s.counted) {
      // drift 窗口
      countedSegments.push({ at: now, ms: s.durationMs });
      while (countedSegments.length > 0 && now - countedSegments[0].at < now - cfg.driftWindowMs) countedSegments.shift();

      // heartbeat：doing 任务连续专注
      const task = getCurrentTask();
      if (task && s.onPrimary) {
        if (continuousTaskId === task.id) {
          continuousFocusMs += s.durationMs;
        } else {
          continuousTaskId = task.id;
          continuousFocusMs = s.durationMs;
        }
      } else {
        resetContinuous();
      }
    } else {
      // 不计入的会话（副屏/空闲）打断连续专注
      resetContinuous();
    }

    evaluate(s.endedAt);
  });

  // 快照变化：idle_back 检测
  onSnapshotChange((snap) => {
    const now = Date.now();
    if (!snap && isIdle()) {
      if (idleStartedAt === 0) idleStartedAt = now;
    } else if (snap && idleStartedAt > 0) {
      const idleDur = now - idleStartedAt;
      idleStartedAt = 0;
      if (idleDur >= cfg.idleBackMinMs && canFire('idle_back', now)) {
        fire('idle_back', `离开了 ${Math.round(idleDur / 60_000)} 分钟。要继续刚才的任务吗？`);
      }
    }
  });
}

function resetContinuous(): void {
  continuousFocusMs = 0;
  continuousTaskId = null;
}

function evaluate(now: number): void {
  // 1. heartbeat
  if (continuousFocusMs >= cfg.heartbeatFocusMs && canFire('heartbeat', now)) {
    const task = getCurrentTask();
    fire('heartbeat', `「${task?.text ?? '当前任务'}」已专注 ${Math.round(continuousFocusMs / 60_000)} 分钟，起来走走。`);
    resetContinuous(); // 提醒后重新累计
    return;
  }

  // 2. switch_storm
  if (switchTimes.length >= cfg.stormSwitchThreshold && canFire('switch_storm', now)) {
    fire('switch_storm', `${cfg.stormWindowMs / 60_000} 分钟内切换了 ${switchTimes.length} 次窗口。先停一下，回到一件事上。`);
    switchTimes.length = 0; // 平复窗口
    return;
  }

  // 3. drift：有 doing 任务但窗口内计入过少
  const task = getCurrentTask();
  if (task) {
    const cutoff = now - cfg.driftWindowMs;
    const counted = countedSegments
      .filter((seg) => seg.at >= cutoff)
      .reduce((sum, seg) => sum + seg.ms, 0);
    if (counted < cfg.driftMinCountedMs && canFire('drift', now)) {
      fire('drift', `「${task.text}」还在等你。刚过去 ${cfg.driftWindowMs / 60_000} 分钟里，主屏只计了 ${Math.round(counted / 1000)} 秒。`);
    }
  }
}

/** 渲染层查询当前提醒配置（M5 设置页） */
export function getReminderConfig(): ReminderConfig {
  return { ...cfg };
}

export function updateReminderConfig(patch: Partial<ReminderConfig>): ReminderConfig {
  cfg = { ...cfg, ...patch };
  return { ...cfg };
}

File: apps/desktop-rs/src-tauri/src/reminders.rs
/**
 * 提醒规则引擎（M4）—— 四条规则 + 去重 + 贪睡静默
 *
 * 与 Electron 版 reminders.ts 对齐（并修正其 drift 窗口剪枝 bug）：
 * 1. heartbeat     同一 doing 任务连续主屏计入 ≥ heartbeatFocusMs → 休息提醒
 * 2. drift         有 doing 任务，但过去 driftWindowMs 主屏计入 < driftMinCountedMs → 轻推
 * 3. switch_storm  stormWindowMs 内前台切换 ≥ stormSwitchThreshold → 平复提醒
 * 4. idle_back     空闲 ≥ idleBackMinMs 后恢复活动 → 欢迎回来
 *
 * 通用约束：
 * - 每种规则 perRuleCooldownMs 内最多触发一次
 * - 「稍后提醒」→ snoozeMs 内静默（针对单条规则或全部）
 * - 触发即发 EVT_REMINDER 事件 → overlay 窗口消费
 * - Tauri 版无 Electron 的独立 overlay 窗口句柄检查，用 OVERLAY_VISIBLE 原子标志代替
 */
use serde::Serialize;
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Mutex;
use tauri::{AppHandle, Emitter};

pub const EVT_REMINDER: &str = "reminders://fired";

/* ── 配置（M5 设置页可改）────────────────────────────────── */

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReminderConfig {
    pub heartbeat_focus_ms: i64,      // 默认 50min
    pub drift_window_ms: i64,         // 默认 5min
    pub drift_min_counted_ms: i64,    // 默认 30s
    pub storm_window_ms: i64,         // 默认 5min
    pub storm_switch_threshold: i64,  // 默认 15
    pub idle_back_min_ms: i64,        // 默认 10min
    pub per_rule_cooldown_ms: i64,    // 默认 10min
    pub snooze_ms: i64,               // 默认 20min
}

impl Default for ReminderConfig {
    fn default() -> Self {
        Self {
            heartbeat_focus_ms: 50 * 60_000,
            drift_window_ms: 5 * 60_000,
            drift_min_counted_ms: 30_000,
            storm_window_ms: 5 * 60_000,
            storm_switch_threshold: 15,
            idle_back_min_ms: 10 * 60_000,
            per_rule_cooldown_ms: 10 * 60_000,
            snooze_ms: 20 * 60_000,
        }
    }
}

type RuleKey = &'static str;

const RULE_TITLE: fn(RuleKey) -> &'static str = |k| match k {
    "heartbeat" => "连续专注了一段时间",
    "drift" => "好像走神了？",
    "switch_storm" => "切换有点频繁",
    _ => "欢迎回来",
};

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OverlayPayload {
    pub kind: String,
    pub title: String,
    pub body: String,
}

/* ── 全局状态 ────────────────────────────────────────────── */

static PAUSED: AtomicBool = AtomicBool::new(false);
static OVERLAY_VISIBLE: AtomicBool = AtomicBool::new(false);

#[derive(Default)]
struct State {
    last_fired_at: HashMap<String, i64>,
    snoozed_until: HashMap<String, i64>,
    // heartbeat：当前 doing 任务连续计入时长（切任务/不计入会话即清零）
    continuous_focus_ms: i64,
    continuous_task_id: Option<i64>,
    // switch_storm：环形切换时刻
    switch_times: Vec<i64>,
    // drift：窗口内计入片段 (at, ms)
    counted_segments: Vec<(i64, i64)>,
}

static STATE: Mutex<Option<State>> = Mutex::new(None);
static CFG: Mutex<Option<ReminderConfig>> = Mutex::new(None);

fn with_state<R>(f: impl FnOnce(&mut State) -> R) -> R {
    let mut guard = STATE.lock().expect("reminders state poisoned");
    let s = guard.get_or_insert_with(State::default);
    f(s)
}

fn config() -> ReminderConfig {
    CFG.lock()
        .expect("reminders cfg poisoned")
        .clone()
        .unwrap_or_default()
}

pub fn current_config() -> ReminderConfig {
    config()
}

pub fn update_config(patch: serde_json::Value) -> ReminderConfig {
    let mut cfg = config();
    if let Some(v) = patch.get("heartbeatFocusMs").and_then(|x| x.as_i64()) {
        cfg.heartbeat_focus_ms = v;
    }
    if let Some(v) = patch.get("driftWindowMs").and_then(|x| x.as_i64()) {
        cfg.drift_window_ms = v;
    }
    if let Some(v) = patch.get("driftMinCountedMs").and_then(|x| x.as_i64()) {
        cfg.drift_min_counted_ms = v;
    }
    if let Some(v) = patch.get("stormWindowMs").and_then(|x| x.as_i64()) {
        cfg.storm_window_ms = v;
    }
    if let Some(v) = patch.get("stormSwitchThreshold").and_then(|x| x.as_i64()) {
        cfg.storm_switch_threshold = v;
    }
    if let Some(v) = patch.get("idleBackMinMs").and_then(|x| x.as_i64()) {
        cfg.idle_back_min_ms = v;
    }
    if let Some(v) = patch.get("perRuleCooldownMs").and_then(|x| x.as_i64()) {
        cfg.per_rule_cooldown_ms = v;
    }
    if let Some(v) = patch.get("snoozeMs").and_then(|x| x.as_i64()) {
        cfg.snooze_ms = v;
    }
    *CFG.lock().expect("reminders cfg poisoned") = Some(cfg.clone());
    cfg
}

pub fn set_paused(p: bool, app: &AppHandle) {
    PAUSED.store(p, Ordering::SeqCst);
    if p {
        hide_overlay(app);
    }
}

pub fn snooze_all(ms: i64) {
    let now = now_ms();
    with_state(|s| {
        for k in ["heartbeat", "drift", "switch_storm", "idle_back"] {
            s.snoozed_until.insert(k.to_string(), now + ms);
        }
    });
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// 设置页「预览提醒」：绕过 canFire 直接触发一条样例
pub fn fire_test(app: &AppHandle, title: &str, body: &str) {
    let payload = OverlayPayload {
        kind: "test".to_string(),
        title: title.to_string(),
        body: body.to_string(),
    };
    show_and_emit(app, &payload);
}

/* ── 浮层窗口句柄（lib.rs 注入）──────────────────────────── */

static OVERLAY_WINDOW: Mutex<Option<tauri::WebviewWindow>> = Mutex::new(None);

pub fn set_overlay_window(win: tauri::WebviewWindow) {
    *OVERLAY_WINDOW.lock().expect("overlay win poisoned") = Some(win);
}

pub fn is_overlay_visible() -> bool {
    OVERLAY_VISIBLE.load(Ordering::SeqCst)
}

pub fn show_overlay() {
    OVERLAY_VISIBLE.store(true, Ordering::SeqCst);
}

pub fn hide_overlay(app: &AppHandle) {
    OVERLAY_VISIBLE.store(false, Ordering::SeqCst);
    crate::overlay::hide_on_main(app);
}

/* ── 触发判定 ────────────────────────────────────────────── */

fn can_fire(rule: RuleKey, now: i64, cfg: &ReminderConfig) -> bool {
    if PAUSED.load(Ordering::SeqCst) {
        return false;
    }
    if is_overlay_visible() {
        return false;
    }
    if now < with_state(|s| s.snoozed_until.get(rule).copied().unwrap_or(0)) {
        return false;
    }
    if now - with_state(|s| s.last_fired_at.get(rule).copied().unwrap_or(0)) < cfg.per_rule_cooldown_ms {
        return false;
    }
    true
}

fn fire(app: &AppHandle, rule: RuleKey, body: String) {
    let now = now_ms();
    with_state(|s| s.last_fired_at.insert(rule.to_string(), now));
    let payload = OverlayPayload {
        kind: rule.to_string(),
        title: RULE_TITLE(rule).to_string(),
        body,
    };
    show_and_emit(app, &payload);
    let _ = app.emit(EVT_REMINDER, &payload);
}

/// 显示浮层 + 25s 自动隐藏（fire / fire_test 共用）
/// 启动初期浮层窗口可能尚未创建完成（异步 build），此处最多等待 5s。
/// 窗口显示/隐藏必须在主线程执行（overlay::show_on_main / hide_on_main）。
fn show_and_emit(app: &AppHandle, payload: &OverlayPayload) {
    let deadline = std::time::Instant::now() + std::time::Duration::from_secs(5);
    loop {
        {
            let guard = OVERLAY_WINDOW.lock().expect("overlay win poisoned");
            if guard.is_some() {
                break;
            }
        }
        if std::time::Instant::now() >= deadline {
            // 浮层始终未就绪：至少把提醒事件发给前端，不留悬挂的可见标志
            let _ = app.emit(EVT_REMINDER, payload);
            return;
        }
        std::thread::sleep(std::time::Duration::from_millis(100));
    }

    OVERLAY_VISIBLE.store(true, Ordering::SeqCst);
    crate::overlay::show_on_main(app);
    let _ = app.emit(EVT_REMINDER, payload);
    let app2 = app.clone();
    std::thread::spawn(move || {
        std::thread::sleep(std::time::Duration::from_millis(25_000));
        if OVERLAY_VISIBLE.load(Ordering::SeqCst) {
            OVERLAY_VISIBLE.store(false, Ordering::SeqCst);
            crate::overlay::hide_on_main(&app2);
        }
    });
}

/* ── 会话事件入口（lib.rs 桥接 tracker::EVT_SESSION）────── */

pub fn on_session_ended(
    app: &AppHandle,
    app_key: &str,
    on_primary: bool,
    counted: bool,
    counted_ms: i64,
) {
    let cfg = config();
    let now = now_ms();
    let _ = app_key;

    with_state(|s| {
        // switch_storm：所有前台会话都算一次切换（含副屏不计入）
        s.switch_times.push(now);
        while !s.switch_times.is_empty() && now - s.switch_times[0] > cfg.storm_window_ms {
            s.switch_times.remove(0);
        }

        if counted {
            // drift 窗口（Electron 版剪枝条件写反，此处修正：保留窗口内片段）
            s.counted_segments.push((now, counted_ms));
            while !s.counted_segments.is_empty() && s.counted_segments[0].0 < now - cfg.drift_window_ms {
                s.counted_segments.remove(0);
            }

            // heartbeat：同一 doing 任务连续主屏计入
            let task = crate::db::current_task().ok().flatten();
            if let Some(t) = task {
                if on_primary {
                    if s.continuous_task_id == Some(t.id) {
                        s.continuous_focus_ms += counted_ms;
                    } else {
                        s.continuous_task_id = Some(t.id);
                        s.continuous_focus_ms = counted_ms;
                    }
                    return; // 注意：闭包内 return 结束本次 with_state
                }
            }
            // 没有 doing 任务 / 非主屏 → 打断连续专注
            s.continuous_focus_ms = 0;
            s.continuous_task_id = None;
        } else {
            // 不计入的会话（副屏/空闲）打断连续专注
            s.continuous_focus_ms = 0;
            s.continuous_task_id = None;
        }
    });

    evaluate(app, now);
}

/// 空闲恢复入口（lib.rs 桥接 tracker::EVT_SNAPSHOT：idle → active）
pub fn on_activity_resumed(app: &AppHandle, idle_ms_before: i64) {
    let cfg = config();
    let now = now_ms();
    if idle_ms_before >= cfg.idle_back_min_ms && can_fire("idle_back", now, &cfg) {
        fire(
            app,
            "idle_back",
            format!("离开了 {} 分钟。要继续刚才的任务吗？", idle_ms_before / 60_000),
        );
    }
}

/// 主循环 tick：评估 drift（有 doing 任务但窗口内计入过少）
pub fn on_minute_tick(app: &AppHandle) {
    evaluate(app, now_ms());
}

fn evaluate(app: &AppHandle, now: i64) {
    let cfg = config();

    // 1. heartbeat
    let (hb_focus, hb_task_id) = with_state(|s| (s.continuous_focus_ms, s.continuous_task_id));
    if hb_focus >= cfg.heartbeat_focus_ms && can_fire("heartbeat", now, &cfg) {
        let task_text = crate::db::list_tasks(true)
            .ok()
            .and_then(|v| {
                let tid = hb_task_id.unwrap_or(-1);
                v.into_iter().find(move |t| t.id == tid)
            })
            .map(|t| t.text)
            .unwrap_or_else(|| "当前任务".to_string());
        fire(
            app,
            "heartbeat",
            format!(
                "「{}」已专注 {} 分钟，起来走走。",
                task_text,
                hb_focus / 60_000
            ),
        );
        with_state(|s| {
            s.continuous_focus_ms = 0;
            s.continuous_task_id = None;
        });
        return;
    }

    // 2. switch_storm
    let storm_n = with_state(|s| s.switch_times.len() as i64);
    if storm_n >= cfg.storm_switch_threshold && can_fire("switch_storm", now, &cfg) {
        fire(
            app,
            "switch_storm",
            format!(
                "{} 分钟内切换了 {} 次窗口。先停一下，回到一件事上。",
                cfg.storm_window_ms / 60_000,
                storm_n
            ),
        );
        with_state(|s| s.switch_times.clear());
        return;
    }

    // 3. drift：有 doing 任务但窗口内计入过少
    if crate::db::current_task().ok().flatten().is_some() {
        let counted_sum = with_state(|s| {
            s.counted_segments
                .iter()
                .filter(|(at, _)| *at >= now - cfg.drift_window_ms)
                .map(|(_, ms)| *ms)
                .sum::<i64>()
        });
        if counted_sum < cfg.drift_min_counted_ms && can_fire("drift", now, &cfg) {
            let task_text = crate::db::current_task()
                .ok()
                .flatten()
                .map(|t| t.text)
                .unwrap_or_else(|| "当前任务".to_string());
            fire(
                app,
                "drift",
                format!(
                    "「{}」还在等你。刚过去 {} 分钟里，主屏只计了 {} 秒。",
                    task_text,
                    cfg.drift_window_ms / 60_000,
                    counted_sum / 1000
                ),
            );
        }
    }
}
